internal/buffer/buffer.go

```diff
@@ -971,29 +976,31 @@ func (b *Buffer) FindMatchingBrace(braceType [2]rune, start Loc) (Loc, bool, boo
 
 // Retab changes all tabs to spaces or vice versa
 func (b *Buffer) Retab() {
-	toSpaces := b.Settings["tabstospaces"].(bool)
-	tabsize := util.IntOpt(b.Settings["tabsize"])
-	dirty := false
+	if !b.Type.Readonly {
```

Comment:
This may be good as a temporary fix. But there is one more issue with Retab: it is not undoable. The reason is that if modifies the buffer directly, not via `EventHandler`, i.e. bypasses the undo stack. So it seems a final solution would be to change Retab to use EventHandler's `Replace` or `MultipleReplace`, and add readonly mode check for any replace operations.